airflow/contrib/hooks/sqoop_hook.py

```diff
@@ -14,210 +14,282 @@
+                                       "Cannot execute {} on {}. Error code is: {}"
+                                       "Output: {}, Stderr: {}"
+                                   ).format(cmd, self.conn.host,
+                                            process.returncode, output,
+                                            stderr))
+
+    def _prepare_command(self, export=False):
         if export:
-            connection_cmd = ["sqoop", "export", "--verbose"]
+            connection_cmd = ["sqoop", "export"]
         else:
-            connection_cmd = ["sqoop", "import", "--verbose"]
+            connection_cmd = ["sqoop", "import"]
 
+        if self.verbose:
+            connection_cmd += ["--verbose"]
         if self.job_tracker:
             connection_cmd += ["-jt", self.job_tracker]
         if self.conn.login:
             connection_cmd += ["--username", self.conn.login]
-        # todo: put this in a password file
         if self.conn.password:
             connection_cmd += ["--password", self.conn.password]
-        if self.lib_jars:
-            connection_cmd += ["-libjars", self.lib_jars]
+        if self.password_file:
+            connection_cmd += ["--password-file", self.password_file]
+        if self.libjars:
+            connection_cmd += ["-libjars", self.libjars]
         if self.files:
             connection_cmd += ["-files", self.files]
         if self.namenode:
             connection_cmd += ["-fs", self.namenode]
         if self.archives:
             connection_cmd += ["-archives", self.archives]
+        if self.num_mappers:
+            connection_cmd += ["--num-mappers", self.num_mappers]
+
+        for key, value in self.properties.items():
+            connection_cmd += ["-D", "{}={}".format(key, value)]
 
-        connection_cmd += ["--connect", "{}:{}/{}".format(self.conn.host, self.conn.port,
-                                                          self.conn.schema)]
-        connection_cmd += cmd
+        connection_cmd += ["--connect", "{}:{}/{}".format(
+            self.conn.host,
+            self.conn.port,
+            self.conn.schema
+        )]
 
         return connection_cmd
 
-    def _import_cmd(self, target_dir,
-                    append=False, type="text",
-                    num_mappers=None, split_by=None):
+    def _get_export_format_argument(self, file_type='text'):
+        if file_type == "avro":
+            return ["--as-avrodatafile"]
+        elif file_type == "sequence":
+            return ["--as-sequencefile"]
+        elif file_type == "parquet":
+            return ["--as-parquetfile"]
+        else:
+            return ["--as-textfile"]
+
+    def _import_cmd(self, target_dir, append, file_type, split_by, direct,
+                    driver):
 
-        cmd = ["--target-dir", target_dir]
+        cmd = self._prepare_command(export=False)
 
-        if not num_mappers:
-            num_mappers = 1
+        cmd += ["--target-dir", target_dir]
```

Comment:
can we make this optional as well? If you pass in a table then you don't have to specify this target-dir, only if you have a query do you need to.

Reply:
Sure, please rewrite in such a way that either one them is required. And write a test :-)

Reply:
after even more review, you don't need it with query either if you're using hcatalog.. so best to keep it optional :) but yes I can add tests! I plan on testing what happens when you pass both a table and a query and catch that case too.